lib/elixir/lib/map.ex

```diff
@@ -371,6 +371,31 @@ defmodule Map do
     :maps.update(key, value, map)
   end
 
+  @doc """
+  Replaces the value under `key` using the given function only if
+  `key` already exists in `map`.
+
+  In comparison to `replace/3`, this can be useful when it's expensive to calculate the value.
+
+  If `key` does not exist, the original map is returned unchanged.
+
+  ## Examples
+
+      iex> Map.replace_lazy(%{a: 1, b: 2}, :a, fn v -> v * 4 end)
+      %{a: 4, b: 2}
+
+      iex> Map.replace_lazy(%{a: 1, b: 2}, :c, fn v -> v * 4 end)
+      %{a: 1, b: 2}
+
+  """
+  @doc since: "1.14.0"
+  @spec replace_lazy(map, key, (existing_value :: value -> new_value :: value)) :: map
+  def replace_lazy(map, key, _) when not is_map_key(map, key), do: map
+
+  def replace_lazy(map, key, fun) when is_function(fun, 1) do
+    Map.replace(map, key, fun.(map[key]))
```

Comment:
Sorry to micro-optimize but this solution does the key lookup twice (or maybe even three times depending on how replace works).

```suggestion
  def replace_lazy(map, key, fun) when is_map(map) and is_function(fun, 1) do
    case map do
      %{^key => val} -> %{map | key => fun.(val)
      %{} -> map
    end
```